Fixes #8532 - Review System.nanoTime() usages.

sbordet opened the pull request:
Introduced `o.e.j.util.NanoTime` class to deal with nanoTimes.

Now NanoTime.now() should be used instead of System.nanoTime(),
and various `<unit>elapsed*()` methods to calculate elapsed times.

Furthermore, comparing 2 nanoTimes should be done via `isBefore()`,
rather than using the `<` operator, which is wrong as specified in
the System.nanoTime() javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #8532 - Review System.nanoTime() usages.

Introduced o.e.j.util.NanoTime class to deal with nanoTimes.

Now NanoTime.now() should be used instead of System.nanoTime(),
and various <unit>elapsed*() methods to calculate elapsed times.

Furthermore, comparing 2 nanoTimes should be done via isBefore(),
rather than using the < operator, which is wrong as specified in
the System.nanoTime() javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/NanoTime.java:
```diff
@@ -0,0 +1,180 @@
+     * @param endNanoTime the end nanoTime
+     * @return the nanoseconds elapsed between the current nanoTime and the given end nanoTime
+     */
+    public static long elapsedTo(long endNanoTime)
```
This name is confusing as `endNanoTime` is supposed to be in the future but the past tense in the method name suggests that it's in the past.
Something like `remaningTo` for instance sounds better to me.

sbordet replied:
I like `remainingTo`, updated.

gregw replied:
`remainingTo` is not good phrasing.  It should be `remainingUntil` or just `until`.   I think using `elapsed`, `since` and `until` is clear.

sbordet committed:
Updates after review, renamed <unit>elapsedTo() to <unit>remainingTo().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw reviewed (changes requested):
Quick review whilst I have network.
Mostly good, but can the methods of NanoTime be renamed:
- elapsed(long, long); since(long); until(long);
- millisElapsed(long, long); millisSince(long); millisUntil(long);
- secondsElapsed(long, long); secondsSince(long); secondsUntil(long);

I'll check more as I go....

gregw commented on jetty-server/src/test/java/org/eclipse/jetty/server/StressTest.java:
```diff
@@ -458,24 +458,19 @@ private static class TestHandler extends HandlerWrapper
+            long delay = NanoTime.millisElapsed(start, now);
             _latencies[2].add(delay);
-            _latencies[3].add((now - start));
+            _latencies[3].add(delay);
```
I don't think this is the same logic?

sbordet replied:
It's not exactly the same logic, but before it was a garbage value, since `baseRequest.getTimeStamp()` returns an epoch millis time, and `delay` was calculated mixing up epoch millis with current nanoTime converted to millis -- a bogus number that could have even be negative.

Maybe `baseRequest.getTimeStamp()` should return a nanoTime, but `CustomRequestLog` relies on it to be a UTC dateTime -- feels like we should either use `Instant` or have 2 values (the nanoTime and the dateTime).

gregw replied:
Ok. It's only a test anyway, so if it is a nonsense value, perhaps delete it.

gregw reviewed (changes requested):
I've looked through all the files and other than the one query noted, I've not seen any issues other than better names for NanoTime methods.

sbordet committed:
Updates after review.
Renamed <unit>elapsedTo() to <unit>until() and
<unit>elapsedFrom() to <unit>since().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (approved):
I like the new names better!
